pyface/data_view/i_data_view_widget.py

```diff
@@ -8,52 +8,341 @@
+
     def _add_event_listeners(self):
+        logger.debug('Adding DataViewWidget listeners')
         super()._add_event_listeners()
         self.observe(
-            self._header_visible_updated, 'header_visible', dispatch='ui')
+            self._header_visible_updated,
+            'header_visible',
+            dispatch='ui',
+        )
+        self.observe(
+            self._selection_type_updated,
+            'selection_type',
+            dispatch='ui',
+        )
+        self.observe(
+            self._selection_mode_updated,
+            'selection_mode',
+            dispatch='ui',
+        )
+        self.observe(
+            self._selection_updated,
+            '_selection.items',
+            dispatch='ui',
+        )
+        if self.control is not None:
+            self._observe_control_selection()
 
     def _remove_event_listeners(self):
+        logger.debug('Removing DataViewWidget listeners')
+        if self.control is not None:
+            self._observe_control_selection(remove=True)
         self.observe(
             self._header_visible_updated,
             'header_visible',
             dispatch='ui',
             remove=True,
         )
+        self.observe(
+            self._selection_type_updated,
+            'selection_type',
+            dispatch='ui',
+            remove=True,
+        )
+        self.observe(
+            self._selection_mode_updated,
+            'selection_mode',
+            dispatch='ui',
+            remove=True,
+        )
+        self.observe(
+            self._selection_updated,
+            '_selection.items',
+            dispatch='ui',
+            remove=True,
+        )
         super()._remove_event_listeners()
 
-    def _header_visible_updated(self, event):
-        """ Observer for header_visible trait. """
-        if self.control is not None:
-            self._set_control_header_visible(event.new)
+    # ------------------------------------------------------------------------
+    # Private methods
+    # ------------------------------------------------------------------------
 
-    def _get_control_header_visible(self):
-        """ Toolkit specific method to get the visibility of the header. """
-        raise NotImplementedError()
+    @contextmanager
+    def _selection_updating(self):
+        """ Context manager to prevent loopback when updating selections. """
+        if self._selection_updating_flag:
+            yield
+        else:
+            self._selection_updating_flag = True
+            try:
+                yield
+            finally:
+                self._selection_updating_flag = False
 
-    def _set_control_header_visible(self, tooltip):
-        """ Toolkit specific method to set the visibility of the header. """
-        raise NotImplementedError()
+    # Trait property handlers
+
+    @cached_property
+    def _get_selection(self):
+        return self._selection
+
+    def _set_selection(self, selection):
```

Comment:
A few branches in this method are untested, and it made the job harder for me when I tried to refactor and create custom traits. This would be a technical debt for similar refactoring in the future. I have already added some tests in #636, it'd be good to bring them in here.

Reply:
I'm currently seeing 100% branch coverage of this method with the current test suite.